### PR TITLE
Feat: Add support for Device Flow grant for Firefly

### PR DESCRIPTION
**Pull request by marcos-albornoz**
This PR provides the support for OAuth 2.0 Device Flow Grant:

- Added the `device-url` flag to provide the device url to request a device code.

Closes https://venafi.atlassian.net/browse/VC-24604

### CODE REVIEW

File: pkg/venafi/firefly/connector.go [hunk 1]
```diff
@@ -114,6 +114,11 @@ func (c *Connector) Authorize(auth *endpoint.Authentication) (token *oauth2.Toke
 		return config.PasswordCredentialsToken(context.Background(), auth.User, auth.Password)
 	}
 
+	// if it's a device flow grant
+	if auth.IdentityProvider.DeviceURL != "" {
+		return c.getDeviceAccessToken(auth)
+	}
+
 	// if it's a client credentials flow grant
 	if auth.ClientSecret != "" {
```
Comment (rvelaVenafi) [inline pkg/venafi/firefly/connector.go:123]: ClientSecret should be the first authorization method we check. As user/password usage is discouraged and device is a niche use case.

File: pkg/venafi/firefly/deviceFlow.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	//setting as default verificationURL the value returned in VerificationURI given that is the used for the most of the
+	// OAuth IdP like Okta, Auth0 and WSO2 return the verification_uri to show to the user
+	verificationURL := devCred.VerificationURI
+	//if that is empty then trying to use the VerificationURL given google uses verification_url to
+	//return the verification_uri to show to the user
+	if verificationURL == "" {
+		verificationURL = devCred.VerificationURL
+	}
+
+	fmt.Printf("please go to %v and enter the code %v . It will expire in %dm and %ds\n\n", verificationURL, devCred.UserCode, devCred.ExpiresIn/60, devCred.ExpiresIn%60)
```
Comment (rvelaVenafi) [inline pkg/venafi/firefly/deviceFlow.go:60]: We should make a bigger and better message here as this is really important the user see it.
One example is how aws-cli does it.

File: pkg/venafi/firefly/deviceFlow.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+		if err != nil {
+			return nil, err
+		}
+
+		//parsing the response
+		token, err := parseWaitingDeviceAuthorizationRequestResult(statusCode, body)
+
+		if err != nil {
+			switch err.Error() {
+			case "authorization_pending":
```
Comment (rvelaVenafi) [inline pkg/venafi/firefly/deviceFlow.go:140]: Make this strings constants

File: pkg/venafi/firefly/deviceFlow.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+				devCred.Interval += 5
+				time.Sleep(time.Duration(devCred.Interval) * time.Second)
+			case "access_denied":
+				return nil, fmt.Errorf("the access from device was denied by the user")
+			case "expired_token":
+				return nil, fmt.Errorf("the device code expired")
+			default:
+				return nil, fmt.Errorf("the authorization failed. %w", err)
+			}
+		} else {
```
Comment (rvelaVenafi) [inline pkg/venafi/firefly/deviceFlow.go:152]: invert this if/else to remove nesting. If err == nil return with token. Then no else is necessary and you can follow the next steps to check the error and the switch.

File: pkg/venafi/firefly/firefly.go [hunk 5]
```diff
@@ -90,20 +92,33 @@ func (c *Connector) GenerateRequest(_ *endpoint.ZoneConfiguration, req *certific
...
 	var payload io.Reader
 	var b []byte
+	var values url.Values
+
+	contentType := "application/json"
+
 	if method == "POST" || method == "PUT" {
-		b, _ = json.Marshal(data)
-		payload = bytes.NewReader(b)
+		//determining if the data is type of url.Values
```
Comment (rvelaVenafi) [inline pkg/venafi/firefly/firefly.go:103]: Do this in a different function to keep code easy to read

**Review by rvelaVenafi — CHANGES_REQUESTED** [review]

File: pkg/venafi/firefly/connector.go (hunk 1 above)
Comment (luispresuelVenafi) [reply pkg/venafi/firefly/connector.go:123]: Level priority:
- ClientSecret
- Device
- User/Password

File: pkg/venafi/firefly/connector.go [hunk 6]
```diff
@@ -114,6 +114,11 @@ func (c *Connector) Authorize(auth *endpoint.Authentication) (token *oauth2.Toke
 		return config.PasswordCredentialsToken(context.Background(), auth.User, auth.Password)
 	}
 
+	// if it's a device flow grant
+	if auth.IdentityProvider.DeviceURL != "" {
+		return c.getDeviceAccessToken(auth)
```
Comment (luispresuelVenafi) [inline pkg/venafi/firefly/connector.go:119]: You can seperate the function from here and verify err, if not error then print log success log and return values:
```
if auth.IdentityProvider.DeviceURL != "" {
                authToken, err = c.getDeviceAccessToken(auth)
                if err != nil {
                   return nil, err
                }
                log.Printf("Successfully authorized device) // Not sure about the log syntax
		return authToken, nil
}
```

File: pkg/venafi/firefly/connector.go (hunk 1 above)
Comment (marcos-albornoz) [reply pkg/venafi/firefly/connector.go]: done

File: pkg/venafi/firefly/deviceFlow.go (hunk 3 above)
Comment (marcos-albornoz) [reply pkg/venafi/firefly/deviceFlow.go]: done

File: pkg/venafi/firefly/deviceFlow.go (hunk 4 above)
Comment (marcos-albornoz) [reply pkg/venafi/firefly/deviceFlow.go]: done

File: pkg/venafi/firefly/deviceFlow.go (hunk 2 above)
Comment (marcos-albornoz) [reply pkg/venafi/firefly/deviceFlow.go]: done

File: pkg/venafi/firefly/connector.go (hunk 6 above)
Comment (marcos-albornoz) [reply pkg/venafi/firefly/connector.go:162]: It's done now into the c.getDeviceAccessToken(auth)

File: pkg/venafi/firefly/firefly.go (hunk 5 above)
Comment (marcos-albornoz) [reply pkg/venafi/firefly/firefly.go:105]: I prefer to keep in this way given here are gotten the payload and the values which only has meaning when the request data is of type url.Values which is expected when the request is for device flow grant given the request is needed to do using form params and these values are stored, if something is wrong, to be used in the build of the error.

**Review by rvelaVenafi — APPROVED** [review]

**Review by luispresuelVenafi — APPROVED** [review]

**Pull request merged**